packages/SwingSet/src/controller/upgradeSwingset.js

```diff
@@ -204,6 +208,156 @@ export const upgradeSwingset = kernelStorage => {
     version = 2;
   }
 
+  if (version < 3) {
+    // v3 means that we've completed remediation for bug #9039
+    console.log(`Starting remediation of bug #9039`);
+
+    // find all terminated vats
+    const terminated = new Set(JSON.parse(getRequired('vats.terminated')));
+
+    // find all live vats
+    const allVatIDs = [];
+    for (const [_name, vatID] of getAllStaticVats(kvStore)) {
+      if (!terminated.has(vatID)) {
+        allVatIDs.push(vatID);
+      }
+    }
+    for (const vatID of getAllDynamicVats(getRequired)) {
+      if (!terminated.has(vatID)) {
+        allVatIDs.push(vatID);
+      }
+    }
+
+    // find all pending notifies
+    const notifies = new Map(); // .get(kpid) = [vatIDs..];
+    const [runHead, runTail] = JSON.parse(getRequired('runQueue'));
+    for (let p = runHead; p < runTail; p += 1) {
+      const rq = JSON.parse(getRequired(`runQueue.${p}`));
+      if (rq.type === 'notify') {
+        const { vatID, kpid } = rq;
+        assert(vatID);
+        assert(kpid);
+        let vats = notifies.get(kpid);
+        if (!vats) {
+          vats = [];
+          notifies.set(kpid, vats);
+        }
+        vats.push(vatID);
+      }
+    }
+    const [accHead, accTail] = JSON.parse(getRequired('acceptanceQueue'));
+    for (let p = accHead; p < accTail; p += 1) {
+      const rq = JSON.parse(getRequired(`acceptanceQueue.${p}`));
+      if (rq.type === 'notify') {
+        const { vatID, kpid } = rq;
+        assert(vatID);
+        assert(kpid);
+        if (!notifies.has(kpid)) {
+          notifies.set(kpid, []);
+        }
+        notifies.get(kpid).push(vatID);
+      }
+    }
+    console.log(` - pending notifies:`, notifies);
+
+    // cache of known-settled kpids: will grow to num(kpids)
+    const settledKPIDs = new Set();
+    const nonSettledKPIDs = new Set();
+    const isSettled = kpid => {
+      if (settledKPIDs.has(kpid)) {
+        return true;
+      }
+      if (nonSettledKPIDs.has(kpid)) {
+        return false;
+      }
+      const state = kvStore.get(`${kpid}.state`);
+      // missing state means the kpid is deleted somehow, shouldn't happen
+      state || Fail`${kpid}.state is missing`;
+      if (state === 'unresolved') {
+        nonSettledKPIDs.add(kpid);
+        return false;
+      }
+      settledKPIDs.add(kpid);
+      return true;
+    };
```

Comment:
```suggestion
    const settlementStatuses = new Map();
    const isSettled = kpid => {
      if (settlementStatuses.has(kpid)) {
        return settlementStatuses.get(kpid);
      }
      const state = kvStore.get(`${kpid}.state`);
      // missing state means the kpid is deleted somehow, shouldn't happen
      state || Fail`${kpid}.state is missing`;
      const result = state !== 'unresolved';
      settlementStatuses.set(kpid, result);
      return result;
    };
```
or if you're willing to extend packages/internal,
```suggestion
    const isSettled = memoizeUnary(kpid => {
      const state = kvStore.get(`${kpid}.state`);
      // missing state means the kpid is deleted somehow, shouldn't happen
      state || Fail`${kpid}.state is missing`;
      return state !== 'unresolved';
    });
```

Reply:
Oh I didn't know we had a memoize in internal now.

Reply:
We don't, it would need to crib from either [notifier](https://github.com/Agoric/agoric-sdk/blob/3007c001ee490c847987bc61d22acf9689611340/packages/notifier/src/publish-kit.js#L66-L84) or [Endo](https://github.com/endojs/endo/blob/4ca4028bb67ba9198060f30a7e5bbff21e8fa27f/packages/memoize/src/memoize.js#L16) and then generalize to support non-weak keys (e.g., as either different function or by supporting an optional `makeCache` parameter in addition to the input function).

Reply:
Well this is actually more of a "makeOnce" store pattern than a memoization, where we can just throw away the store after usage.

Reply:
I like the Map-based cache, I'll do that. I don't want to modify other packages in this PR if I can avoid it.